sklearn/datasets/_svmlight_format_fast.pyx

```diff
@@ -116,3 +116,146 @@ def _load_svmlight_file(f, dtype, bint multilabel, bint zero_based,
+    cython.floating
+    signed long long
+
+ctypedef fused int_or_float2:
+    cython.integral
+    cython.floating
+    signed long long
+
+ctypedef fused int_or_longlong:
+    cython.integral
+    signed long long
+
+def _dump_svmlight_file_dense(int_or_float1[:,:] X, int_or_float2[:,:] y, f, bint multilabel, bint one_based, int_or_longlong[:] query_id):
+    if int_or_float1 in cython.floating:
+        value_pattern = "%d:%.16g"
+    else:
+        value_pattern = "%d:%d"
+    if int_or_float2 in cython.floating:
+        label_pattern = "%.16g"
+    else:
+        label_pattern = "%d"
+
+    line_pattern = "%s"
+    if query_id is not None:
+        line_pattern += " qid:%d"
+    line_pattern += " %s\n"
+
+    cdef Py_ssize_t num_labels = y.shape[1]
+    cdef Py_ssize_t x_len = X.shape[0]
+    cdef Py_ssize_t row_length = X.shape[1]
+    cdef Py_ssize_t i
+    cdef Py_ssize_t j
+    cdef Py_ssize_t col_start
+    cdef Py_ssize_t col_end
+    cdef bint first
+    for i in range(x_len):
+        s = ""
+        first = True
+        for j in range(row_length):
+            val = X[i,j]
+            if val != 0:
+                if not first:
+                    s += " "
+                first = False
+                s += value_pattern % (j+one_based, val)
+
+        if multilabel:
+            labels_str = ""
+            first = True
+            for j in range(num_labels):
+                val = y[i,j]
+                if val != 0:
+                    if not first:
+                        labels_str += ","
+                    first = False
+                    labels_str += label_pattern % j
+        else:
+            labels_str = label_pattern % y[i,0]
+
+        if query_id is not None:
+            feat = (labels_str, query_id[i], s)
+        else:
+            feat = (labels_str, s)
+
+        f.write((line_pattern % feat).encode("ascii"))
+
+def _dump_svmlight_file_general(X, y, f, bint multilabel, bint one_based, np.ndarray query_id, bint X_is_sp, bint y_is_sp):
+    if X.dtype.kind == "i":
+        value_pattern = "%d:%d"
+    else:
+        value_pattern = "%d:%.16g"
+    if y.dtype.kind == "i":
+        label_pattern = "%d"
+    else:
+        label_pattern = "%.16g"
+
+    line_pattern = "%s"
+    if query_id is not None:
+        line_pattern += " qid:%d"
+    line_pattern += " %s\n"
+
+    cdef Py_ssize_t num_labels = y.shape[1]
+    cdef Py_ssize_t x_len = X.shape[0]
+    cdef Py_ssize_t row_length = X.shape[1]
+    cdef Py_ssize_t i
+    cdef Py_ssize_t j
+    cdef Py_ssize_t col_start
+    cdef Py_ssize_t col_end
+    cdef bint first
+    for i in range(x_len):
+        s = ""
+        first = True
+        if X_is_sp:
+            col_start = X.indptr[i]
+            col_end = X.indptr[i+1]
+            for j in range(col_start, col_end):
+                if not first:
+                    s += " "
+                first = False
+                s += value_pattern % (X.indices[j] + one_based, X.data[j])
```

Comment:
Based on your benchmarks, `main`'s approach is faster than the PR for the sparse case, so let's inherit main's approach here.